src/main/java/org/javarosa/core/model/ItemsetBinding.java

```diff
@@ -200,53 +201,60 @@ private SelectChoice getChoiceForTreeReference(FormDef formDef, DataInstance for
-                    currentAnswersInNewChoices.put(selection.choice != null ? selection.choice.getValue() : selection.xmlValue, false);
+                    currentAnswersInNewChoices.put(selection.choice != null ? selection.choice.getValue() : selection.xmlValue, null);
                 }
             } else {
-                currentAnswersInNewChoices.put(rawValue.getDisplayText(), false);
+                currentAnswersInNewChoices.put(rawValue.getDisplayText(), null);
             }
         }
 
         return currentAnswersInNewChoices;
     }
 
-    private void updateQuestionAnswerInModel(FormDef formDef, TreeReference curQRef, Map<String, Boolean> currentAnswersInNewChoices) {
-        IAnswerData rawValue = formDef.getMainInstance().resolveReference(curQRef).getValue();
+    private void updateQuestionAnswerInModel(FormDef formDef, TreeReference curQRef, Map<String, SelectChoice> currentAnswersInNewChoices) {
+        IAnswerData originalValue = formDef.getMainInstance().resolveReference(curQRef).getValue();
 
-        if (currentAnswersInNewChoices != null && currentAnswersInNewChoices.containsValue(false)) {
-            IAnswerData filteredAnswer;
-            if (rawValue instanceof MultipleItemsData) {
-                filteredAnswer = getFilteredSelections((MultipleItemsData) rawValue, currentAnswersInNewChoices);
+        if (currentAnswersInNewChoices != null) {
+            IAnswerData boundAndFilteredValue;
+            if (originalValue instanceof MultipleItemsData) {
+                boundAndFilteredValue = getFilteredAndBoundSelections((MultipleItemsData) originalValue, currentAnswersInNewChoices);
+            } else if (currentAnswersInNewChoices.containsValue(null)) {
+                boundAndFilteredValue = new StringData("");
             } else {
-                filteredAnswer = new StringData("");
+                SelectChoice selectChoice = currentAnswersInNewChoices.get(originalValue.getDisplayText());
+                boundAndFilteredValue = new SelectOneData(selectChoice.selection());
             }
 
-            formDef.getMainInstance().resolveReference(curQRef).setAnswer(filteredAnswer);
+            formDef.getMainInstance().resolveReference(curQRef).setAnswer(boundAndFilteredValue);
         }
     }
 
     /**
      * @param selections          an answer to a multiple selection question
-     * @param shouldKeepSelection maps each value that could be in @{code selections} to a boolean representing whether
-     *                            or not it should be kept
-     * @return a copy of {@code selections} without the values that were mapped to false in {@code shouldKeepSelection}
+     * @param selectChoicesToKeep maps each value that could be in @{code selections} to a SelectChoice if it should be bound
+     *                            or null if it should be removed.
+     * @return a copy of {@code selections} without the values that were mapped to false in {@code selectChoicesToKeep} and
```

Comment:
Should the comment still say “mapped to false”?

Reply:
What a pleasant way to say I made a mistake. 😂